pkg/api_builder/lib/api_builder.dart

```diff
@@ -0,0 +1,99 @@
+import 'dart:convert';
+import 'package:shelf/shelf.dart' as shelf;
+
+export 'package:shelf_router/shelf_router.dart' show Router;
+export 'package:shelf/shelf.dart' show Request, Response;
+
+/// Annotation for an API end-point.
+class EndPoint {
+  /// HTTP verb for requests routed to the annotated method.
+  final String verb;
+
+  /// HTTP route for request routed to the annotated method.
+  final String route;
+
+  /// Create an annotation that routes requests matching [verb] and [route] to
+  /// the annotated method.
+  const EndPoint(this.verb, this.route);
+
+  /// Route `GET` requests matching [route] to annotated method.
+  const EndPoint.get(this.route) : verb = 'GET';
+
+  /// Route `HEAD` requests matching [route] to annotated method.
+  const EndPoint.head(this.route) : verb = 'HEAD';
+
+  /// Route `POST` requests matching [route] to annotated method.
+  const EndPoint.post(this.route) : verb = 'POST';
+
+  /// Route `PUT` requests matching [route] to annotated method.
+  const EndPoint.put(this.route) : verb = 'PUT';
+
+  /// Route `DELETE` requests matching [route] to annotated method.
+  const EndPoint.delete(this.route) : verb = 'DELETE';
+
+  /// Route `CONNECT` requests matching [route] to annotated method.
+  const EndPoint.connect(this.route) : verb = 'CONNECT';
+
+  /// Route `OPTIONS` requests matching [route] to annotated method.
+  const EndPoint.options(this.route) : verb = 'OPTIONS';
+
+  /// Route `TRACE` requests matching [route] to annotated method.
+  const EndPoint.trace(this.route) : verb = 'TRACE';
+}
+
+/// Thrown to cause a different response than what the end-point would normally
+/// return.
+class ResponseException implements Exception {
+  final int status;
+  final List<int> body;
```

Comment:
I'm wondering: would it be better if `RequestException` and/or `ResponseException` could accept any body, and then callers of this would not need to manually encode the json.

On the other hand, if it is only this utility that emits them, maybe we could auto-decode them to `Map` or `String` (if we know of the type).

Reply:
I was thinking we would just make subclasses of `ResponseException`, so we would make the `AuthorizationException extends RequestException` somewhere in our code..

But `RequestException` is only thrown by the client. I don't think we can do anything better there..

Reply:
Actually, we could have subclasses if `RequestException` in `api_builder`..

Note: `AuthorizationException` is something I would define in `app/shared/exceptions.dart` or something like that...